### PR TITLE
[Merged by Bors] - feat (LinearAlgebra/RootPairing): define RootPairingCat

### PR DESCRIPTION
**Pull request by ScottCarnahan**
This PR defines a category of root pairings, roughly following the treatment of morphisms of root data in SGA3 Exp.21 Sec.6.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/RootSystem/RootPairingCat.lean [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+Authors: Scott Carnahan
+-/
+import Mathlib.LinearAlgebra.RootSystem.Defs
+import Mathlib.CategoryTheory.Category.Basic
+
+/-!
+# The category of root pairings
+This file defines the category of root pairings, following the definition of category of root data
+given in SGA III Exp. 21 Section 6.
+## Main definitions:
```
Comment (ocfnash) [inline Mathlib/LinearAlgebra/RootSystem/RootPairingCat.lean:13]:
```suggestion

## Main definitions:
```

File: Mathlib/LinearAlgebra/RootSystem/RootPairingCat.lean [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  weight_map : P.weight →ₗ[R] Q.weight
+  /-- A contravariant linear map on coweight space. -/
+  coweight_map : Q.coweight →ₗ[R] P.coweight
+  /-- A bijection on index sets. -/
+  index_map : P.index ≃ Q.index
+  weight_coweight_transpose :
+    LinearMap.dualMap weight_map =
+      P.pairing.toDualRight ∘ₗ coweight_map ∘ₗ Q.pairing.toDualRight.symm
+  root_weight_map : weight_map ∘ P.pairing.root = Q.pairing.root ∘ index_map
+  coroot_coweight_map : coweight_map ∘ Q.pairing.coroot = P.pairing.coroot ∘ index_map.symm
```
Comment (ocfnash) [inline Mathlib/LinearAlgebra/RootSystem/RootPairingCat.lean:75]: I think we should first define morphisms without using the category theory library.

For example we could create a new file (maybe called `RootSystem/Hom.lean`) and in it define:
```lean
structure RootPairing.Hom (ι₂ M₂ N₂ : Type*)
    [AddCommGroup M₂] [Module R M₂] [AddCommGroup N₂] [Module R N₂]
    (P : RootPairing ι R M N) (Q : RootPairing ι₂ R M₂ N₂) where
  /-- A linear map on weight space. -/
  weightMap : M →ₗ[R] M₂
  /-- A contravariant linear map on coweight space. -/
  coweightMap : N₂ →ₗ[R] N
  /-- A bijection on index sets. -/
  indexEquiv : ι ≃ ι₂
  weight_coweight_transpose : weightMap.dualMap ∘ₗ Q.toDualRight = P.toDualRight ∘ₗ coweightMap
  root_weightMap : weightMap ∘ P.root = Q.root ∘ indexEquiv
  coroot_coweightMap : coweightMap ∘ Q.coroot = P.coroot ∘ indexEquiv.symm
```
together with basic boilerplate and then use this to define the `Category` instance below (similar to what is done for the category of modules [here](https://github.com/leanprover-community/mathlib4/blob/00df099f8ed42c12b35cdedcfbf1fde6c7413662/Mathlib/Algebra/Category/ModuleCat/Basic.lean#L91-L97)).

I claim the advantages of this design are:

- We can have universe polymorphic morphisms
- It is possible to speak of morphisms without having to invoke the category theory library
- It is more idiomatic compared to the rest of our linear algebra

What do you think?

File: Mathlib/LinearAlgebra/PerfectPairing.lean [hunk 3]
```diff
@@ -49,6 +49,9 @@ instance instFunLike : FunLike (PerfectPairing R M N) M (N →ₗ[R] R) where
   coe f := f.toLin
   coe_injective' x y h := by cases x; cases y; simpa using h
 
+lemma toLin_apply (p : PerfectPairing R M N) {x : M} : p.toLin x = p x := by
```
[the remark this answers is not in the document]
Comment (ScottCarnahan) [reply Mathlib/LinearAlgebra/PerfectPairing.lean:53]: Thank you for this suggestion.  I had tried it earlier and got intimidated by the amount of rearranging that became necessary.  I'm glad I tried again.

[the remark this answers is not in the document]
Comment (ocfnash) [reply Mathlib/LinearAlgebra/PerfectPairing.lean:53]: Thanks so much for doing this!

File: Mathlib/LinearAlgebra/RootSystem/RootPairingCat.lean (hunk 2 above)
Comment (ocfnash) [reply Mathlib/LinearAlgebra/RootSystem/RootPairingCat.lean]: No need IMHO, though I'm excited to see what comes next :)

**Comment by ocfnash** [comment]
Thanks so much for developing all this root system theory!

bors merge

**Pull request closed without merging**